src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/ILGeneratorImpl.cs

```diff
@@ -398,7 +535,40 @@ private static int GetStackChange(OpCode opcode, MethodInfo methodInfo, Type[]?
 
         public override void EmitCalli(OpCode opcode, CallingConventions callingConvention, Type? returnType, Type[]? parameterTypes, Type[]? optionalParameterTypes) => throw new NotImplementedException();
         public override void EmitCalli(OpCode opcode, CallingConvention unmanagedCallConv, Type? returnType, Type[]? parameterTypes) => throw new NotImplementedException();
-        public override void EndExceptionBlock() => throw new NotImplementedException();
+
+        public override void EndExceptionBlock()
+        {
+            if (_currentExStackIndex <= -1)
+            {
+                throw new NotSupportedException(SR.Argument_NotInExceptionBlock);
+            }
+
+            ExceptionBlock currentExBlock = _exceptionStack[_currentExStackIndex];
+            ExceptionState state = currentExBlock.State;
+            Label endLabel = currentExBlock.EndLabel;
+
+            if (state == ExceptionState.Filter || state == ExceptionState.Try)
+            {
+                throw new InvalidOperationException(SR.Argument_BadExceptionCodeGen);
+            }
+
+            if (state == ExceptionState.Catch)
+            {
+                Emit(OpCodes.Leave, endLabel);
+                MarkLabel(currentExBlock.HandleEnd);
+            }
+            else if (state == ExceptionState.Finally || state == ExceptionState.Fault)
+            {
+                Emit(OpCodes.Endfinally);
+                MarkLabel(currentExBlock.HandleEnd);
+            }
+
+            MarkLabel(endLabel);
+            currentExBlock.State = ExceptionState.Done;
+            _exceptionStack.Remove(currentExBlock);
+            _currentExStackIndex--;
```

Comment:
During the removal, is there any state we should assert? Some internal state that we are tracking as opposed to what the user is expected to do.

Reply:
Could assert the removed index is the last element in the list

Reply:
> Could assert the removed index is the last element in the list

Well now when using `_exceptionStack.Count` instead of tracking `_currentExStackIndex` separately there is no need to assert that, if that is what you meant.